[SPARK-21213][SQL] Support collecting partition-level statistics: rowCount and sizeInBytes

mbasmanova opened the pull request:
## What changes were proposed in this pull request?

Added support for ANALYZE TABLE [db_name].tablename PARTITION (partcol1[=val1], partcol2[=val2], ...) COMPUTE STATISTICS [NOSCAN] SQL command to calculate total number of rows and size in bytes for a subset of partitions. Calculated statistics are stored in Hive Metastore as user-defined properties attached to partition objects. Property names are the same as the ones used to store table-level statistics: spark.sql.statistics.totalSize and spark.sql.statistics.numRows.

When partition specification contains all partition columns with values, the command collects statistics for a single partition that matches the specification. When some partition columns are missing or listed without their values, the command collects statistics for all partitions which match a subset of partition column values specified.

For example, table t has 4 partitions with the following specs:

* Partition1: (ds='2008-04-08', hr=11)
* Partition2: (ds='2008-04-08', hr=12)
* Partition3: (ds='2008-04-09', hr=11)
* Partition4: (ds='2008-04-09', hr=12)

'ANALYZE TABLE t PARTITION (ds='2008-04-09', hr=11)' command will collect statistics only for partition 3.

'ANALYZE TABLE t PARTITION (ds='2008-04-09')' command will collect statistics for partitions 3 and 4.

'ANALYZE TABLE t PARTITION (ds, hr)' command will collect statistics for all four partitions.

When the optional parameter NOSCAN is specified, the command doesn't count number of rows and only gathers size in bytes.

The statistics gathered by ANALYZE TABLE command can be fetched using DESC EXTENDED [db_name.]tablename PARTITION command.

## How was this patch tested?

Added tests.

mbasmanova committed:
[SPARK-21213][SQL] Support collecting partition-level statistics: rowCount and sizeInBytes

AmplabJenkins commented:
Can one of the admins verify this patch?

gatorsmile commented:
ok to test

gatorsmile commented:
add to whitelist

gatorsmile commented:
cc @wzhfy

SparkQA commented:
**[Test build #78663 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78663/testReport)** for PR 18421 at commit [`e18f144`](https://github.com/apache/spark/commit/e18f144a2c6b318a285fbb184bb5c78408f06039).

SparkQA commented:
**[Test build #78663 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78663/testReport)** for PR 18421 at commit [`e18f144`](https://github.com/apache/spark/commit/e18f144a2c6b318a285fbb184bb5c78408f06039).
 * This patch **fails PySpark pip packaging tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78663/
Test FAILed.

gatorsmile commented:
Please update the PR title. 
> [SPARK-21213][SQL] Support collecting partition-level statistics: rowCount and sizeInBytes

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala:
```diff
@@ -96,7 +96,8 @@ object CatalogStorageFormat {
     storage: CatalogStorageFormat,
-    parameters: Map[String, String] = Map.empty) {
+    parameters: Map[String, String] = Map.empty,
+    stats: Option[CatalogStatistics] = None) {
```
Add this part to the description of this case class.

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala:
```diff
@@ -101,19 +101,30 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder(conf) {
+
+    val table = visitTableIdentifier(ctx.tableIdentifier)
+    if (ctx.identifierSeq() == null) {
+      AnalyzeTableCommand(table, noscan, partitionSpec)
```
`AnalyzeTableCommand(table, noscan = ctx.identifier != null, partitionSpec)`

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala:
```diff
@@ -101,19 +101,30 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder(conf) {
+      true
+    } else {
+      false
+    }
```
```Scala
    if (ctx.identifier != null &&
        ctx.identifier.getText.toLowerCase(Locale.ROOT) != "noscan") {
      throw new ParseException(s"Expected `NOSCAN` instead of `${ctx.identifier.getText}`", ctx)
    }
```

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala:
```diff
@@ -101,19 +101,30 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder(conf) {
+      Option(ctx.partitionSpec).map(visitNonOptionalPartitionSpec)
     } else {
+      None
+    }
```
```val partitionSpec = Option(ctx.partitionSpec).map(visitNonOptionalPartitionSpec)```

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala:
```diff
@@ -101,19 +101,30 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder(conf) {
    * }}}
```
Please also update the above description with newly supported the partition spec.

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeTableCommand.scala:
```diff
@@ -24,18 +24,23 @@ import scala.util.control.NonFatal
 /**
- * Analyzes the given table to generate statistics, which will be used in query optimizations.
+ * Analyzes the given table or partition to generate statistics, which will be used in
+ * query optimizations.
```
Could you please add the description about `partitionSpec`? 
>  If certain partition specs are specified, then statistics are gathered for only those partitions.

gatorsmile commented on sql/core/src/test/scala/org/apache/spark/sql/execution/SparkSqlParserSuite.scala:
```diff
@@ -239,18 +239,20 @@ class SparkSqlParserSuite extends AnalysisTest {
-      AnalyzeTableCommand(TableIdentifier("t"), noscan = true))
+      AnalyzeTableCommand(TableIdentifier("t"), noscan = true,
+        partitionSpec = Some(Map("ds" -> "2008-04-09", "hr" -> "11"))))
+    intercept("ANALYZE TABLE t PARTITION(ds, hr) COMPUTE STATISTICS",
```
This should be legal based on the description of Hive? 

https://cwiki.apache.org/confluence/display/Hive/StatsDev

mbasmanova replied:
I see. Thanks for pointing that out. Currently, this PR supports only exact partition spec. It doesn't support partial partition specs describing the document above. My preference would be to keep it simple for this PR and support only exact spec and add support for partial specs in a follow up PR. What do you think?

gatorsmile replied:
I am fine about this. Please write a comment in your PR description, the class description of `visitAnalyze` and `AnalyzeTableCommand `

wzhfy replied:
If we first supports only exact partition spec, in the near future we will change (replace) the syntax and also the related implementation. I mean we are not doing it incrementally, so I prefer support the right syntax in this single pr. Would support that syntax require lots of work?

mbasmanova replied:
@wzhfy, I expect syntax change to be quite small and incremental. Currently, it is necessary to specify all partition columns along with values. The change will be to allow only a subset of partition columns and allow partition columns without values.

`PARTITION (partcol1=val1,...) -> PARTITION (partcol1[=val1],...)`

That said, I want to try to allow partial partition specs in this PR. Let me spend some time on it and report back my findings.

mbasmanova committed:
[SPARK-21213][SQL] review comments

SparkQA commented:
**[Test build #78749 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78749/testReport)** for PR 18421 at commit [`f9327b0`](https://github.com/apache/spark/commit/f9327b0105c42e4eeaf8f5be2b9ff51e30c4c892).

SparkQA commented:
**[Test build #78750 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78750/testReport)** for PR 18421 at commit [`090be78`](https://github.com/apache/spark/commit/090be78f016ad8ca213725f16229e43a10c9c71c).

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala:
```diff
@@ -95,25 +95,32 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder(conf) {
    * }}}
+   * Example SQL for analyzing a single partition :
+   * {{{
+   *   ANALYZE TABLE table PARTITION (key=value,..) COMPUTE STATISTICS [NOSCAN];
```
The existing syntax is not good. Could you improve them?
```SQL
ANALYZE TABLE [db_name.]tablename [PARTITION(partcol1[=val1], partcol2[=val2], ...)] 
COMPUTE STATISTICS [NOSCAN]
```

In addition, since we have a restriction, please do not call `visitNonOptionalPartitionSpec`. Instead, we can capture the non-set partition column and issue a more user-friendly exception message.

mbasmanova replied:
I'm seeing that visitNonOptionalPartitionSpec detects unset partition columns and throws an exception: Found an empty partition key '$key'. Is this sufficient or do you have something else in mind?

/**
   * Create a partition specification map without optional values.
   */
  protected def visitNonOptionalPartitionSpec(
      ctx: PartitionSpecContext): Map[String, String] = withOrigin(ctx) {
    visitPartitionSpec(ctx).map {
      case (key, None) => throw new ParseException(s"Found an empty partition key '$key'.", ctx)
      case (key, Some(value)) => key -> value
    }
  }

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala:
```diff
@@ -95,25 +95,32 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder(conf) {
    * {{{
    *   ANALYZE TABLE table COMPUTE STATISTICS [NOSCAN];
```
Also here.

SparkQA commented:
**[Test build #78750 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78750/testReport)** for PR 18421 at commit [`090be78`](https://github.com/apache/spark/commit/090be78f016ad8ca213725f16229e43a10c9c71c).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78750/
Test FAILed.

SparkQA commented:
**[Test build #78749 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78749/testReport)** for PR 18421 at commit [`f9327b0`](https://github.com/apache/spark/commit/f9327b0105c42e4eeaf8f5be2b9ff51e30c4c892).
 * This patch **fails due to an unknown error code, -9**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78749/
Test FAILed.

mbasmanova committed:
[SPARK-21213][SQL] improved comments per review feedback

mbasmanova committed:
[SPARK-21213][SQL] typo

SparkQA commented:
**[Test build #78809 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78809/testReport)** for PR 18421 at commit [`2d1e817`](https://github.com/apache/spark/commit/2d1e817e9a3d214f89eb9389e64b444ea6d38f91).

SparkQA commented:
**[Test build #78809 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78809/testReport)** for PR 18421 at commit [`2d1e817`](https://github.com/apache/spark/commit/2d1e817e9a3d214f89eb9389e64b444ea6d38f91).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #78923 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78923/testReport)** for PR 18421 at commit [`70b35ce`](https://github.com/apache/spark/commit/70b35ce127081a6a125dac223637abec0425ac17).

mbasmanova committed:
[SPARK-21213][SQL] add support for partial partition specs

SparkQA commented:
**[Test build #78927 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78927/testReport)** for PR 18421 at commit [`aecb5a9`](https://github.com/apache/spark/commit/aecb5a908e9c07de9209a7154c48c1262d0ee7f5).

mbasmanova commented:
@wzhfy, @gatorsmile, I updated this PR to support partial partition specs where values are defined only for a subset of partition columns. For example, if table has 2 partition columns ds and hr, both PARTITION (ds='2010-01-01', hr=10) and PARTITION (hr=10) specs are valid. I did not provide support for partition specs where partition column is mentioned without any value. E.g. PARTITION (ds, hr=10) spec is still not allowed. I didn't implement this because I realized that I don't understand why this syntax is useful. My understanding is that (ds, hr=10) is equivalent to (hr=10). Do you think it is important to implement this syntax? If so, would you help me understand why? Is this because we want to support HQL at the same level as Hive to allow seamless transition from Hive to Spark?

SparkQA commented:
**[Test build #78923 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78923/testReport)** for PR 18421 at commit [`70b35ce`](https://github.com/apache/spark/commit/70b35ce127081a6a125dac223637abec0425ac17).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78923/
Test PASSed.

mbasmanova committed:
[SPARK-21213][SQL] add support for partition specs where some partition columns specified with no values

mbasmanova committed:
[SPARK-21213][SQL] comment update

SparkQA commented:
**[Test build #78929 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78929/testReport)** for PR 18421 at commit [`b0804c5`](https://github.com/apache/spark/commit/b0804c513e04ff6480c8633eb26c4f29ffb23e64).

mbasmanova committed:
[SPARK-21213][SQL] removed extra space

mbasmanova commented:
@wzhfy, @gatorsmile, I updated PR to add full support for partition partition specs. This version supports a subset of partition columns with or without values specified. I'm going away for the long weekend tomorrow and will be back next Wed (July 5). I may not be able to respond to comments or provide updates to this PR until then.

SparkQA commented:
**[Test build #78930 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78930/testReport)** for PR 18421 at commit [`b527f18`](https://github.com/apache/spark/commit/b527f186fe85c8b8117ac9dbd3a62b5ff2e2f473).

SparkQA commented:
**[Test build #78927 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78927/testReport)** for PR 18421 at commit [`aecb5a9`](https://github.com/apache/spark/commit/aecb5a908e9c07de9209a7154c48c1262d0ee7f5).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78927/
Test FAILed.

SparkQA commented:
**[Test build #78929 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78929/testReport)** for PR 18421 at commit [`b0804c5`](https://github.com/apache/spark/commit/b0804c513e04ff6480c8633eb26c4f29ffb23e64).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78929/
Test PASSed.

SparkQA commented:
**[Test build #78930 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78930/testReport)** for PR 18421 at commit [`b527f18`](https://github.com/apache/spark/commit/b527f186fe85c8b8117ac9dbd3a62b5ff2e2f473).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78930/
Test PASSed.

mbasmanova commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala:
```diff
@@ -986,6 +987,7 @@ private[hive] object HiveClientImpl {
     tpart.setTableName(ht.getTableName)
     tpart.setValues(partValues.asJava)
     tpart.setSd(storageDesc)
+    tpart.setParameters(mutable.Map(p.parameters.toSeq: _*).asJava)
```
p.parameters.asJava produces as immutable map and causes java.lang.UnsupportedOperationException: remove;

mbasmanova committed:
[SPARK-21213][SQL] addressed easy review comments

SparkQA commented:
**[Test build #79232 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79232/testReport)** for PR 18421 at commit [`0150fea`](https://github.com/apache/spark/commit/0150fea7e0f159e7411a1779cfa5cefc765e4a9c).

SparkQA commented:
**[Test build #79232 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79232/testReport)** for PR 18421 at commit [`0150fea`](https://github.com/apache/spark/commit/0150fea7e0f159e7411a1779cfa5cefc765e4a9c).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79232/
Test FAILed.

mbasmanova commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala:
```diff
@@ -201,6 +202,193 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
+      assert(queryStats(ds, hr) === None)
+    }
+
+    def createPartition(ds: String, hr: Int, query: String): Unit = {
```
I created this method to make the test code more concise. For example, this method helps avoid multi-line methods calls like this:

sql(
        s"""
           |INSERT INTO TABLE $tableName PARTITION (ds='2010-01-02', hr=11)
           |SELECT * FROM src
           |UNION ALL
           |SELECT * FROM src
         """.stripMargin)

becomes 

createPartition("2010-01-02", 11, "SELECT * FROM SRC UNION ALL SELECT * FROM SRC")

mbasmanova committed:
[SPARK-21213][SQL] addressed remaining review comments

mbasmanova commented:
@wzhfy, thank you for review. I created AnalyzePartitionCommand class, modified the logic to update all partitions in a single call to Metastore, combined noscan and non-noscan test cases. I believe I addressed all comments except for requests to make queryStats and assertStats common functions. After merging test cases, each of these functions is used only in one test, hence, there is no copy-pasta anymore. Would you take another look?

SparkQA commented:
**[Test build #79246 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79246/testReport)** for PR 18421 at commit [`2d8732e`](https://github.com/apache/spark/commit/2d8732e27892cc79196cae63815b2d500d506975).

mbasmanova commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala:
```diff
@@ -201,6 +202,193 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
+    }
+  }
+
+  test("analyze a set of partitions") {
```
@wzhfy, given that these cases are covered in SparkSqlParserSuite, is it still necessary to cover them again in StatisticsSuite? Partition columns without values are removed at parsing stage so that AnalyzePartitionCommand always receives partition column along with a value.

re: (ds, hr='10') - https://cwiki.apache.org/confluence/display/Hive/StatsDev suggests that it is allowed; this PR supports this syntax.

wzhfy replied:
@mbasmanova IIUC, the logic is wrong here. For example, when analyzing partition (ds, hr), we should not remove them in parser. Currently we parse it to `AnalyzeTableCommand`, which collects table-level stats. But what we need to do is to collect **partition-level stats for all partitions**.
Check hive's behavior [here](https://cwiki.apache.org/confluence/display/Hive/StatsDev#StatsDev-Examples)

mbasmanova committed:
[SPARK-21213][SQL] added test case for (ds, hr=11) partition spec

SparkQA commented:
**[Test build #79248 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79248/testReport)** for PR 18421 at commit [`8c0b61b`](https://github.com/apache/spark/commit/8c0b61bd693cd2fb5a6e3a691261455fdc27dfcc).

SparkQA commented:
**[Test build #79248 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79248/testReport)** for PR 18421 at commit [`8c0b61b`](https://github.com/apache/spark/commit/8c0b61bd693cd2fb5a6e3a691261455fdc27dfcc).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79248/
Test FAILed.

SparkQA commented:
**[Test build #79246 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79246/testReport)** for PR 18421 at commit [`2d8732e`](https://github.com/apache/spark/commit/2d8732e27892cc79196cae63815b2d500d506975).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class AnalyzePartitionCommand(`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79246/
Test PASSed.

SparkQA commented:
**[Test build #79285 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79285/testReport)** for PR 18421 at commit [`43f2112`](https://github.com/apache/spark/commit/43f2112116257d95431fd3aed2cf832580410902).

SparkQA commented:
**[Test build #79285 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79285/testReport)** for PR 18421 at commit [`43f2112`](https://github.com/apache/spark/commit/43f2112116257d95431fd3aed2cf832580410902).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79285/
Test PASSed.

wzhfy commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala:
```diff
@@ -93,30 +93,50 @@ class SparkSqlAstBuilder extends AstBuilder {
-      AnalyzeTableCommand(visitTableIdentifier(ctx.tableIdentifier), noscan = false)
     } else {
+      if (partitionSpec.isDefined) {
+        logWarning(s"Partition specification is ignored: ${ctx.partitionSpec.getText}")
```
Partition specification is ignored when collecting column statistics

wzhfy commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala:
```diff
@@ -93,30 +93,50 @@ class SparkSqlAstBuilder extends AstBuilder {
-    } else if (ctx.identifierSeq() == null) {
-      AnalyzeTableCommand(visitTableIdentifier(ctx.tableIdentifier), noscan = false)
     } else {
+      if (partitionSpec.isDefined) {
```
should use `if (ctx.partitionSpec != null)`, currently we don't support partition-level column stats.

wzhfy commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala:
```diff
@@ -93,30 +93,50 @@ class SparkSqlAstBuilder extends AstBuilder {
-      if (ctx.identifier.getText.toLowerCase(Locale.ROOT) != "noscan") {
-        throw new ParseException(s"Expected `NOSCAN` instead of `${ctx.identifier.getText}`", ctx)
+
+    val partitionSpec =
```
`partitionSpec` ->`partitionValueSpec`

wzhfy commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeTableCommand.scala:
```diff
@@ -17,14 +17,17 @@
 /**
- * Analyzes the given table to generate statistics, which will be used in query optimizations.
+ * Analyzes the given table to generate statistics, which will be used in
+ * query optimizations.
```
unnecessary change

wzhfy commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeTableCommand.scala:
```diff
@@ -72,3 +61,81 @@ case class AnalyzeTableCommand(
+ * Analyzes a given set of partitions to generate per-partition statistics, which will be used in
+ * query optimizations.
+ */
+case class AnalyzePartitionCommand(
```
shall we create a new file for this class?

wzhfy commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala:
```diff
@@ -181,6 +182,151 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
+      assert(stats.sizeInBytes === sizeInBytes)
+    }
+
+    def assertSizeInBytesStats(ds: String, hr: String, sizeInBytes: BigInt): Unit = {
```
We can merge this method with `assertStats` by adding `rowCount: Option[BigInt]` as a paramter.
Also please rename `assertStats` to `assertPartitionStats`.

wzhfy commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala:
```diff
@@ -181,6 +182,151 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
+    }
+  }
+
+  test("analyzing views is not supported") {
```
I think this is caused by rebase? please remove it.

wzhfy commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala:
```diff
@@ -181,6 +182,151 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
+
+      createPartition("2010-01-01", SELECT_FROM_SRC)
+      createPartition("2010-01-02", s"$SELECT_FROM_SRC UNION ALL $SELECT_FROM_SRC")
+      createPartition("2010-01-03", SELECT_FROM_SRC)
```
Since it's a test for `analyze single partition`, we only need one partition here, the others seems redundant.

mbasmanova replied:
@wzhfy, I'm using multiple partitions to protect against bugs like the following:

- ANALYZE PARTITION collects stats for the whole table instead of a single partition (hence, test table needs to have 2+ partitions)
- ANALYZE PARTITION collects stats for the *first* partition instead of the partition specified in the SQL command (hence, need to analyze a couple of different partitions)

Using "analyze single partition" for the test name I was hoping to communicate the intent of testing analyzing of a single partition as opposed to a set of partitions. Does this make sense? Do you still want me to reduce the test to a single-partition table?

wzhfy commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala:
```diff
@@ -181,6 +182,151 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
     }
   }
 
+  private val SELECT_FROM_SRC = "SELECT '1', 'A' from src"
```
we can just use `SELECT '1', 'A'` and remove this val.

wzhfy commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala:
```diff
@@ -181,6 +182,151 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
+      assertSizeInBytesStats("2010-01-01", "10", 2000)
+      assertSizeInBytesStats("2010-01-01", "11", 2000)
+      assertSizeInBytesStats("2010-01-02", "10", 2000)
+      assertSizeInBytesStats("2010-01-02", "11", 2*2000)
```
Checks for `ds='2010-01-02'` has no difference with `ds='2010-01-01'`, we can remove this.

mbasmanova replied:
I've added checks for all 4 partitions after *every* analyze statement to make sure there are no unintended side effects, e.g. analyzing partition A doesn't change stats for partition B. Would it be ok to keep this logic or do you still think I should remove it?

wzhfy commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala:
```diff
@@ -181,6 +182,151 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
+      assertStats("2010-01-01", "10", 500, 2000)
+      assertStats("2010-01-01", "11", 500, 2000)
+      assertStats("2010-01-02", "10", 500, 2000)
+      assertStats("2010-01-02", "11", 2*500, 2*2000)
```
same here, we can remove this.

wzhfy commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala:
```diff
@@ -181,6 +182,151 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
+
+      sql(s"ANALYZE TABLE $tableName PARTITION (ds='2010-01-01') COMPUTE STATISTICS")
+
+      assertStats("2010-01-01", "10", 500, 2000)
```
`assertStats("2010-01-01", "10", rowCount = 500, sizeInBytes = 2000)`

wzhfy commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala:
```diff
@@ -93,30 +93,50 @@ class SparkSqlAstBuilder extends AstBuilder {
+      if (partitionSpec.isDefined) {
+        AnalyzePartitionCommand(table, partitionSpec.get, noscan = ctx.identifier != null)
+      } else {
+        AnalyzeTableCommand(table, noscan = ctx.identifier != null)
```
@mbasmanova IIUC, the logic is wrong here. For example, when analyzing partition (ds, hr), we should not remove them in parser. Currently we parse it to AnalyzeTableCommand, which collects table-level stats. But what we need to do is to collect partition-level stats for all partitions.
Check hive's behavior here

mbasmanova replied:
@wzhfy , good catch. Thank you! I'll fix this.

mbasmanova committed:
[SPARK-21213][SQL] addressed review comments; fixed PARTITION (ds, hr) case

SparkQA commented:
**[Test build #79526 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79526/testReport)** for PR 18421 at commit [`32edbe8`](https://github.com/apache/spark/commit/32edbe8c92ed47ac7eeb45546aa21866a486c89a).

mbasmanova committed:
[SPARK-21213][SQL] shorted new test

SparkQA commented:
**[Test build #79529 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79529/testReport)** for PR 18421 at commit [`fd819be`](https://github.com/apache/spark/commit/fd819be32c0257d48667561db5da39829659a570).

mbasmanova committed:
[SPARK-21213][SQL] added documentation; added test for an empty table

mbasmanova commented:
@wzhfy, I fixed the case of PARTITION (ds, hr) to collect stats for *all* partitions instead of for a table as a whole. Rebased. Addressed other comments. Please, take another look.

SparkQA commented:
**[Test build #79533 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79533/testReport)** for PR 18421 at commit [`994cd06`](https://github.com/apache/spark/commit/994cd06526d78e56f3d8a9c823d93d097f91fb6b).

SparkQA commented:
**[Test build #79529 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79529/testReport)** for PR 18421 at commit [`fd819be`](https://github.com/apache/spark/commit/fd819be32c0257d48667561db5da39829659a570).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79529/
Test FAILed.

SparkQA commented:
**[Test build #79533 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79533/testReport)** for PR 18421 at commit [`994cd06`](https://github.com/apache/spark/commit/994cd06526d78e56f3d8a9c823d93d097f91fb6b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79533/
Test FAILed.

SparkQA commented:
**[Test build #79526 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79526/testReport)** for PR 18421 at commit [`32edbe8`](https://github.com/apache/spark/commit/32edbe8c92ed47ac7eeb45546aa21866a486c89a).
 * This patch passes all tests.
 * This patch **does not merge cleanly**.
 * This patch adds the following public classes _(experimental)_:
  * `case class AnalyzePartitionCommand(`

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79526/
Test PASSed.

mbasmanova commented:
@wzhfy, @gatorsmile, I'm on vacation till the end of July. I hope this PR is good to merge now, but if not, I'll address further comments first week of August.

mbasmanova commented:
Jenkins, retest this please

SparkQA commented:
**[Test build #79729 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79729/testReport)** for PR 18421 at commit [`994cd06`](https://github.com/apache/spark/commit/994cd06526d78e56f3d8a9c823d93d097f91fb6b).

mbasmanova commented:
@wzhfy, @gatorsmile, I just noticed that Jenkins reported test failures. I'll look into these first week of August and update this PR.

gatorsmile commented:
Will review it tonight. Thanks!

SparkQA commented:
**[Test build #79729 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79729/testReport)** for PR 18421 at commit [`994cd06`](https://github.com/apache/spark/commit/994cd06526d78e56f3d8a9c823d93d097f91fb6b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79729/
Test FAILed.

gatorsmile commented:
Is that possible the partition-level row counts is larger than the table-level row counts after running this new command?

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala:
```diff
@@ -112,4 +112,29 @@ object CommandUtils extends Logging {
+    val oldTotalSize = oldStats.map(_.sizeInBytes.toLong).getOrElse(0L)
+    val oldRowCount = oldStats.flatMap(_.rowCount.map(_.toLong)).getOrElse(-1L)
+    var newStats: Option[CatalogStatistics] = None
+    if (newTotalSize > 0 && newTotalSize != oldTotalSize) {
```
Is that possible the new one is equal to zero?

mbasmanova replied:
Indeed. I fixed this to use -1 and >=0.

gatorsmile commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala:
```diff
@@ -1028,25 +994,115 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
+      table: CatalogTable): CatalogTablePartition = {
+    val restoredSpec = restorePartitionSpec(partition.spec, table.partitionColumnNames)
+
+    // construct Spark's statistics from information in Hive metastore
```
Here, we only respect Spark own statistics. Please also clarify it here.

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala:
```diff
@@ -112,4 +112,29 @@ object CommandUtils extends Logging {
+      oldStats: Option[CatalogStatistics],
+      newTotalSize: BigInt,
+      newRowCount: Option[BigInt]): Option[CatalogStatistics] = {
+    val oldTotalSize = oldStats.map(_.sizeInBytes.toLong).getOrElse(0L)
```
The latest code has a change here. Update the initial value to `-1L`

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeTableCommand.scala:
```diff
@@ -37,31 +37,17 @@ case class AnalyzeTableCommand(
       }
-    }
-    // Update the metastore if the above statistics of the table are different from those
-    // recorded in the metastore.
```
add it back?

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala:
```diff
@@ -90,30 +90,40 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder(conf) {
-   * Example SQL for analyzing table :
+   * Create an [[AnalyzeTableCommand]] command, or an [[AnalyzePartitionCommand]]
+   * or an [[AnalyzeColumnCommand]] command.
+   * Example SQL for analyzing table or a set of partitions :
```
`table or a set of partitions :` -> `a table or a set of partitions:`

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala:
```diff
@@ -90,30 +90,40 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder(conf) {
     } else {
+      if (ctx.partitionSpec != null) {
+        logWarning("Partition specification is ignored when collecting column statistics: " +
+          ctx.partitionSpec.getText)
```
Why?

gatorsmile replied:
We should issue an exception here; otherwise, users might get confused since it has different behaviors.

mbasmanova replied:
This is the existing behavoir. ANALYZE TABLE command simply ignores PARTITION clause. This PR is adding support for PARTITION clause for COMPUTE STATISTICS, but keeps COMPUTE STATISTICS FOR COLUMNS as is. I'm planning to add partition support to FOR COLUMN In a follow-up PR.

Changing this code to throw an exception in this PR will *break* existing uses if folks are relying on a WARN.

Would it be OK to keep this functionality unchanged in the PR?

gatorsmile replied:
Ok. Let us keep it unchanged,

gatorsmile commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala:
```diff
@@ -182,6 +183,189 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
+        sql(s"ANALYZE TABLE $tableName PARTITION (hour=20) COMPUTE STATISTICS")
+      }
+
+      intercept[AnalysisException] {
```
Could you capture all these error messages? We always need to check whether the error messages make sense.
```Scala
      val message = intercept[AnalysisException] {
        ...
      }.getMessage
      assert(message.contains("XYZ"))
```

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzePartitionCommand.scala:
```diff
@@ -0,0 +1,149 @@
+      }
+      Some(filters.reduce(And))
+    } else {
+      None
```
Change it to `Literal.TrueLiteral`

mbasmanova replied:
@gatorsmile , I don't understand this comment. Do you want me to change None to Literal.TrueLiteral? This doesn't compile and doesn't make sense to me as Literal.TrueLiteral just represents 'true' which is not a valid filter. Could you explain a little bit?

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzePartitionCommand.scala:
```diff
@@ -0,0 +1,149 @@
+      tableDf.filter(Column(filter.get)).groupBy(partitionColumns: _*).count()
+    } else {
+      tableDf.groupBy(partitionColumns: _*).count()
+    }
```
After the above change, then this line can be simplified to 
```Scala
val df = tableDf.filter(Column(filter)).groupBy(partitionColumns: _*).count()
```

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzePartitionCommand.scala:
```diff
@@ -0,0 +1,149 @@
+      tableDf.groupBy(partitionColumns: _*).count()
+    }
+
+    val numPartitionColumns = partitionColumns.size
```
Please Inline this one

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzePartitionCommand.scala:
```diff
@@ -0,0 +1,149 @@
+    df.collect().map { r =>
+      val partitionColumnValues = partitionColumns.indices.map(r.get(_).toString)
+      val spec: TablePartitionSpec =
+        tableMeta.partitionColumnNames.zip(partitionColumnValues).toMap
```
Nit: `val spec = tableMeta.partitionColumnNames.zip(partitionColumnValues).toMap`

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzePartitionCommand.scala:
```diff
@@ -0,0 +1,149 @@
+    // recorded in the metastore.
+    val partitionStats = partitions.map { p =>
+      val newTotalSize = CommandUtils.calculateLocationSize(sessionState,
+        tableMeta.identifier, p.storage.locationUri)
```
Nit: 
```
      val newTotalSize = CommandUtils.calculateLocationSize(
        sessionState, tableMeta.identifier, p.storage.locationUri)
```

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzePartitionCommand.scala:
```diff
@@ -0,0 +1,149 @@
+
+    val newPartitions = partitionStats.filter(_._2.isDefined).map { case (p, newStats) =>
+      p.copy(stats = newStats)
+    }.toList
```
Line 98-108 can be simplified to
```Scala
    val newPartitions = partitions.flatMap { p =>
      val newTotalSize = CommandUtils.calculateLocationSize(
        sessionState, tableMeta.identifier, p.storage.locationUri)
      val newRowCount = rowCounts.get(p.spec)
      val newStats = CommandUtils.compareAndGetNewStats(tableMeta.stats, newTotalSize, newRowCount)
      newStats.map(_ => p.copy(stats = newStats))
    }
```

gatorsmile commented:
ping @mbasmanova

mbasmanova commented:
@gatorsmile, I was on vacation and today is my first day back. I'm planning to work through the feedback on this PR today.

mbasmanova commented:
[ @gatorsmile ] Is that possible the partition-level row counts is larger than the table-level row counts after running this new command?

I think so. If table contents have changed since table-level stats were collected and a new very large partition was added or existing one replaced, collecting partition-level stats for that one partition will result in a partition row-count exceeding table row-count.

I think you are pointing out to a larger issue of staleness of statistics which results from table modifications. My thinking is that ANALYZE TABLE commands provide the necessary primitives to collect statistics, but a separate facility is needed to manage the process of collecting these stats. At a very basic level a user would be required to manually run necessary ANALYZE TABLE commands when table content is changed, but for larger deployments an automated solution could be devised.

mbasmanova committed:
[SPARK-21213][SQL] review comments

SparkQA commented:
**[Test build #80082 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80082/testReport)** for PR 18421 at commit [`a0fd270`](https://github.com/apache/spark/commit/a0fd270154e09de862ed1fd04c765c5f5256fdce).

SparkQA commented:
**[Test build #80082 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80082/testReport)** for PR 18421 at commit [`a0fd270`](https://github.com/apache/spark/commit/a0fd270154e09de862ed1fd04c765c5f5256fdce).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/80082/
Test PASSed.

mbasmanova commented:
@gatorsmile , I believe I addressed all your comments and I'm seeing that tests are passing now. Thanks for pointers to -1 and <=0 changes. Tests were failing before because I missed these changes. Let me know if anything else needs more work or if this is now good to merge.

mbasmanova committed:
[SPARK-21213][SQL] fixed bad merge of SPARK-21599

SparkQA commented:
**[Test build #80353 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80353/testReport)** for PR 18421 at commit [`2233b2c`](https://github.com/apache/spark/commit/2233b2c602862c1382107cc817d46744c039571c).

gatorsmile commented:
Sorry, I missed this. Will review it today.

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala:
```diff
@@ -107,6 +109,7 @@ case class CatalogTablePartition(
     if (parameters.nonEmpty) {
       map.put("Partition Parameters", s"{${parameters.map(p => p._1 + "=" + p._2).mkString(", ")}}")
     }
+    stats.foreach(s => map.put("Partition Statistics", s.simpleString))
```
This needs a test case. Could you add a new test file to [this test suite](https://github.com/apache/spark/tree/4ecc648ad713f9d618adf0406b5d39981779059d/sql/core/src/test/resources/sql-tests/inputs) for analyzing and desc table partitions? You can generate the result file by running the command:

> SPARK_GENERATE_GOLDEN_FILES=1 build/sbt "sql/test-only *SQLQueryTestSuite"

mbasmanova replied:
@gatorsmile , this code doesn't make partition stats appear in the output of DESC command. It only adds stats into to CatalogTablePartition.toString output (similar to CatalogTable.toString). Do you still want me to add some tests for this functionality?

gatorsmile replied:
Yes, please add it. Try it and we should expose it to the external users.

mbasmanova replied:
Indeed. It works like you said it would. :-) Adding a test.

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzePartitionCommand.scala:
```diff
@@ -0,0 +1,138 @@
+ *
+ * If `partitionSpec` mentions unknown partition column, an `AnalysisException` is raised.
+ *
+ * By default, total number of rows and total size in bytes is calculated. When `noscan`
```
Nit: `is` -> `are`

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzePartitionCommand.scala:
```diff
@@ -0,0 +1,138 @@
+ * If `partitionSpec` mentions unknown partition column, an `AnalysisException` is raised.
+ *
+ * By default, total number of rows and total size in bytes is calculated. When `noscan`
+ * is `false`, only total size in bytes is computed.
```
`false` -> `true`

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzePartitionCommand.scala:
```diff
@@ -0,0 +1,138 @@
+
+  private def validatePartitionSpec(table: CatalogTable): Option[TablePartitionSpec] = {
+    val partitionColumnNames = table.partitionColumnNames.toSet
+    val invalidColumnNames = partitionSpec.keys.filterNot(partitionColumnNames.contains(_))
```
Here, we also need to consider the conf `spark.sql.caseSensitive`. There are multiple examples for the similar issues.

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzePartitionCommand.scala:
```diff
@@ -0,0 +1,138 @@
+    partitionSpec: Map[String, Option[String]],
+    noscan: Boolean = true) extends RunnableCommand {
+
+  private def validatePartitionSpec(table: CatalogTable): Option[TablePartitionSpec] = {
```
`validatePartitionSpec ` -> `getDynamicPartitionSpec`?

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzePartitionCommand.scala:
```diff
@@ -0,0 +1,138 @@
+        invalidColumnNames.mkString(","))
+    }
+
+    val filteredSpec = partitionSpec.filter(_._2.isDefined)
```
In the table insert with the mixed partition spec, we have the following requirement:
> In part_spec, the static partition keys must come before the dynamic partition keys. That means, all partition columns having constant values need to appear before other partition columns that do not have an assigned constant value.

Do we have similar requirement in analyze table partitions?

mbasmanova replied:
@gatorsmile , this PR doesn't impose this requirement. Hive docs also doesn't list it: https://cwiki.apache.org/confluence/display/Hive/StatsDev

gatorsmile commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala:
```diff
@@ -256,6 +257,201 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
+      assertPartitionStats("2010-01-01", "11", rowCount = Some(500), sizeInBytes = 2000)
+      assertPartitionStats("2010-01-02", "10", rowCount = Some(500), sizeInBytes = 2000)
+      assertPartitionStats("2010-01-02", "11", rowCount = Some(2*500), sizeInBytes = 2*2000)
+    }
```
Does the following query work in Hive?

```Scala
sql(s"ANALYZE TABLE $tableName PARTITION (hr = 10) COMPUTE STATISTICS")
```

mbasmanova replied:
I can't tell from the Hive docs: https://cwiki.apache.org/confluence/display/Hive/StatsDev

Will try it.

gatorsmile replied:
Thanks!

mbasmanova replied:
@gatorsmile , the installation of Hive I have access fails the above query:

> Partition columns in partition specification are not the same as that defined in the table schema. The names and orders have to be exactly the same. Partition columns in the table schema are: (ds, hr), while the partitions specified in the query are: (hr).

It also fails any partition specification which doesn't list all partition columns. E.g., all of (ds), (hr), (ds='2017-01-01'), (hr=10) fail with the above error message.

This installation is not using the latest open source Hive though.

SparkQA commented:
**[Test build #80353 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80353/testReport)** for PR 18421 at commit [`2233b2c`](https://github.com/apache/spark/commit/2233b2c602862c1382107cc817d46744c039571c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeTableCommand.scala:
```diff
@@ -37,31 +37,19 @@ case class AnalyzeTableCommand(
+      if (noscan) {
+        None
+      } else {
+        Some(BigInt(sparkSession.table(tableIdentWithDB).count()))
```
Let us shorten it to a single line.
```Scala
    val newRowCount =
      if (noscan) None else Some(BigInt(sparkSession.table(tableIdentWithDB).count()))
```

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/80353/
Test PASSed.

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeTableCommand.scala:
```diff
@@ -19,7 +19,7 @@ package org.apache.spark.sql.execution.command
 import org.apache.spark.sql.{AnalysisException, Row, SparkSession}
 import org.apache.spark.sql.catalyst.TableIdentifier
-import org.apache.spark.sql.catalyst.catalog.{CatalogStatistics, CatalogTableType}
+import org.apache.spark.sql.catalyst.catalog.{CatalogTableType}
```
Nit: `import org.apache.spark.sql.catalyst.catalog.CatalogTableType`

gatorsmile commented:
@mbasmanova Could you also please update the PR description to describe the user APIs and their semantics? Thanks!

mbasmanova committed:
[SPARK-21213][SQL] added support for spark.sql.caseSensitive; addressed other review comments

SparkQA commented:
**[Test build #80401 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80401/testReport)** for PR 18421 at commit [`64efad2`](https://github.com/apache/spark/commit/64efad20eaa5253a08cd69f89e11550fd0246187).

mbasmanova committed:
[SPARK-21213][SQL] addressed remaining review comments

SparkQA commented:
**[Test build #80402 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80402/testReport)** for PR 18421 at commit [`ba48073`](https://github.com/apache/spark/commit/ba4807342db2943d72c32e62a0bf043efeb573a3).

SparkQA commented:
**[Test build #80401 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80401/testReport)** for PR 18421 at commit [`64efad2`](https://github.com/apache/spark/commit/64efad20eaa5253a08cd69f89e11550fd0246187).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/80401/
Test PASSed.

SparkQA commented:
**[Test build #80402 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80402/testReport)** for PR 18421 at commit [`ba48073`](https://github.com/apache/spark/commit/ba4807342db2943d72c32e62a0bf043efeb573a3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/80402/
Test PASSed.

mbasmanova commented:
@gatorsmile, thanks for review. Comments addressed. Questions answered. Tests as passing. Would you take another look?

gatorsmile commented:
There are still two remaining comments. Could you also address them?

gatorsmile commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala:
```diff
@@ -1033,25 +998,126 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
+      table: CatalogTable): CatalogTablePartition = {
+    val restoredSpec = restorePartitionSpec(partition.spec, table.partitionColumnNames)
+
+    // Restore Spark's statistics from information in Metastore.
```
Also leave comments here to explain the partition-level stats are unavailable prior to 2.3

gatorsmile commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala:
```diff
@@ -256,6 +257,222 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
+          sql(s"ANALYZE TABLE $tableName PARTITION (DS='2010-01-01') COMPUTE STATISTICS")
+        }.getMessage
+        assert(message.contains(s"Partition specification for table '${tableName.toLowerCase}' " +
+            "in database 'default' refers to unknown partition column(s): DS"))
```
Nit: Indent.

gatorsmile commented:
@mbasmanova Really thank you for your contribution! LGTM except the above four comments.

mbasmanova committed:
[SPARK-21213][SQL] Added a test for DESC PARTITION after ANALYZE; review comments

SparkQA commented:
**[Test build #80492 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80492/testReport)** for PR 18421 at commit [`54ec425`](https://github.com/apache/spark/commit/54ec425956c1d1026372e66c8b74496f0615dabf).

gatorsmile commented:
This looks great! 

Regarding this https://github.com/apache/spark/pull/18421/files#r132493566, could we just follow Hive? That is consistent with what we are doing in INSERT. 

In the future releases, we still can make a change if needed. Thanks again!

SparkQA commented:
**[Test build #80492 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80492/testReport)** for PR 18421 at commit [`54ec425`](https://github.com/apache/spark/commit/54ec425956c1d1026372e66c8b74496f0615dabf).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/80492/
Test FAILed.

mbasmanova committed:
[SPARK-21213][SQL] added DROP TABLE to describe-part-after-analyze.sql

SparkQA commented:
**[Test build #80501 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80501/testReport)** for PR 18421 at commit [`b1e5079`](https://github.com/apache/spark/commit/b1e5079812e7411bc00f23e859877bf52792e122).

SparkQA commented:
**[Test build #80501 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80501/testReport)** for PR 18421 at commit [`b1e5079`](https://github.com/apache/spark/commit/b1e5079812e7411bc00f23e859877bf52792e122).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/80501/
Test PASSed.

mbasmanova commented:
@gatorsmile , thank you for reviewing this PR. I'm exciting that it is getting very close to be merged. 

It is not obvious to me why would you prefer to follow Hive and restrict the user to partitions spec which contains partition columns from the head of the list of all partition columns specified in the table schema and disallow arbitrary subsets.

mbasmanova commented:
ping @gatorsmile

gatorsmile commented:
This is just to make it consistent with the partition spec in our current INSERT statement. Could you justify why we need to make them inconsistent? Thanks! 

Also cc @sameeragarwal

mbasmanova committed:
[SPARK-21213][SQL] check that partition columns in the partition spec form a prefix of the partition columns defined in table schema

SparkQA commented:
**[Test build #80806 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80806/testReport)** for PR 18421 at commit [`87594d6`](https://github.com/apache/spark/commit/87594d6aee3e7e1d5a3e5e340331dd7965ea5426).

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzePartitionCommand.scala:
```diff
@@ -0,0 +1,156 @@
+
+  private def getPartitionSpec(table: CatalogTable): Option[TablePartitionSpec] = {
+    val partitionColumnNames = table.partitionColumnNames.toSet
+    val partitionSpecWithCase =
```
Instead of changing the case, could you call `sparkSession.sessionState.conf.resolver`? There are many examples in the code base. Thanks!

SparkQA commented:
**[Test build #80806 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80806/testReport)** for PR 18421 at commit [`87594d6`](https://github.com/apache/spark/commit/87594d6aee3e7e1d5a3e5e340331dd7965ea5426).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/80806/
Test PASSed.

mbasmanova committed:
[SPARK-21213][SQL] use PartitioningUtils.normalizePartitionSpec to handle conf.caseSensitiveAnalysis

SparkQA commented:
**[Test build #80816 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80816/testReport)** for PR 18421 at commit [`3353afa`](https://github.com/apache/spark/commit/3353afadde5ba684372d2ea16ef38ccf22d59d09).

SparkQA commented:
**[Test build #80816 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80816/testReport)** for PR 18421 at commit [`3353afa`](https://github.com/apache/spark/commit/3353afadde5ba684372d2ea16ef38ccf22d59d09).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/80816/
Test FAILed.

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzePartitionCommand.scala:
```diff
@@ -0,0 +1,149 @@
+
+    // Report an error if partition columns in partition specification do not form
+    // a prefix of the list of partition columns defined in the table schema
+    val isSpecified =
```
-> `isNotSpecified`

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzePartitionCommand.scala:
```diff
@@ -0,0 +1,149 @@
+  private def getPartitionSpec(table: CatalogTable): Option[TablePartitionSpec] = {
+    val normalizedPartitionSpec =
+      PartitioningUtils.normalizePartitionSpec(partitionSpec, table.partitionColumnNames,
+        table.identifier.quotedString, conf.resolver);
```
Nit: remove `;`

mbasmanova committed:
[SPARK-21213][SQL] review comments

SparkQA commented:
**[Test build #80846 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80846/testReport)** for PR 18421 at commit [`8ffb140`](https://github.com/apache/spark/commit/8ffb140eb76dd1c41a1c16b2f657cfc0e1ed653c).

SparkQA commented:
**[Test build #80846 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80846/testReport)** for PR 18421 at commit [`8ffb140`](https://github.com/apache/spark/commit/8ffb140eb76dd1c41a1c16b2f657cfc0e1ed653c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/80846/
Test PASSed.

gatorsmile commented:
Thanks!!! Merging to master.

mbasmanova commented:
@gatorsmile , so excited to see this change merged. Thank you for your careful detailed reviews.

wzhfy commented:
@mbasmanova Great work!  I was really busy in the past two months so I didn't have time to look at this.
Thank @gatorsmile for reviewing and merging this PR!

gatorsmile commented:
@wzhfy Welcome back!
